apps/yggdrasil/test/yggdrasil/user_role.exs

```diff
@@ -0,0 +1,104 @@
+defmodule Yggdrasil.UserRoleTest do
+  use ExUnit.Case, async: false
+
+  alias Yggdrasil.{Repo, User, Role, UserRole, RoleResource, Resource, Permission}
+
+  @user %{username: "tester",
+          password: "password",
+          password_confirmation: "password"}
+
+  @test_perms [:foo_perm, :bar_perm, :baz_perm]
+  @test_resources [:foo_res, :bar_res, :baz_res]
+  @test_roles [
+    %{
+      name: :foo_role,
+      resources: [
+        %{
+          name: :foo_res,
+          perms: [:foo_perm, :bar_perm]
+        },
+        %{
+          name: :bar_res,
+          perms: [:bar_perm, :baz_perm]
+        }
+      ]
+    }
+  ]
+
+  setup tags do
+    unless tags[:async] do
+      Ecto.Adapters.SQL.restart_test_transaction(Yggdrasil.Repo, [])
+    end
+
+    :ok
+  end
+
+  setup _ctx do
+    perms = Enum.map @test_perms, fn p ->
+      Repo.insert! %Permission{name: Atom.to_string(p)}
+    end
+
+    res = Enum.map @test_resources, fn r ->
+      Repo.insert! %Resource{name: Atom.to_string(r)}
+    end
+
+    roles = Enum.map @test_roles, fn tr ->
+      role = Repo.insert! %Role{name: Atom.to_string(tr.name)}
+
+      Enum.each tr.resources, fn trs ->
+        resource = Enum.find res, fn rs -> rs.name == Atom.to_string(trs.name) end
+
+        Enum.each trs.perms, fn trpm ->
+          permission = Enum.find perms, fn pm -> 
+            pm.name == Atom.to_string(trpm)
+          end
+
+          Repo.insert! %RoleResource{
+            role_id: role.id,
+            resource_id: resource.id,
```

Comment:
If you made the resource name and permission name their primary key, then inserting this would be pretty easy. No need to lookup, nor would you have to fetch them when you query. The table would make sure only valid resources and permissions were added but wouldn't be needed in the app code.